File: orchestrator/src/storage/redis/scripts/select_instance.lua
-- SPDX-FileCopyrightText: OpenTalk GmbH <mail@opentalk.eu>
--
-- SPDX-License-Identifier: EUPL-1.2

local resource_lookup_key = KEYS[1]
local service_kind_key = KEYS[2]

local resource_value = ARGV[1]

local service_id = redis.call("GET", resource_lookup_key)

if service_id then
    -- found a service that manages the given resource
    local api_key_ids = redis.call("HGET", "ot-orchestrator:service:" .. service_id, "api_key_ids");
    local metrics =  redis.call("HMGET", "ot-orchestrator:service:" .. service_id .. ":metrics", "load", "accepting_jobs")

    return {service_id, api_key_ids, metrics}
end

-- select instance with lowest load
local services = redis.call("SMEMBERS", service_kind_key)

local lowest_load = tonumber("101")
local lowest_load_instance = nil
local lowest_load_metrics = nil
local lowest_load_instance_resource_count = nil

for i, service_id in ipairs(services) do
    local metrics = redis.call("HMGET", "ot-orchestrator:service:" .. service_id .. ":metrics", "load", "accepting_jobs")
    local resource_count = redis.call("SCARD", "ot-orchestrator:service:" .. service_id .. ":resources")

    if metrics ~= nil then
        local load = tonumber(metrics[1])
        local accepting_jobs = tonumber(metrics[2])

        if accepting_jobs == 1 then
            if load < lowest_load
                or (load == lowest_load and resource_count < lowest_load_instance_resource_count) then
                lowest_load = load
                lowest_load_instance = service_id
                lowest_load_metrics = metrics
                lowest_load_instance_resource_count = resource_count
            end
        end
    end
end

if not lowest_load_instance then
    return 1 -- no instance available
end

-- add resource to the lowest load instance
local service_id = lowest_load_instance

redis.call("SADD", "ot-orchestrator:service:".. service_id .. ":resources", resource_value)
redis.call("SET", resource_lookup_key, service_id)

local api_key_ids = redis.call("HGET", "ot-orchestrator:service:" .. service_id, "api_key_ids");

return {service_id, api_key_ids, lowest_load_metrics}

File: orchestrator/src/storage/redis/scripts/select_instance.rs
// SPDX-FileCopyrightText: OpenTalk GmbH <mail@opentalk.eu>
//
// SPDX-License-Identifier: EUPL-1.2

use anyhow::Context;
use base64::{Engine as _, engine::general_purpose};
use opentalk_orchestrator_shared::{Metrics, ServiceKind};
use opentalk_service_auth::ApiKeyId;
use redis::FromRedisValue;
use url::Url;

use crate::storage::{
    InstanceData, ServiceResource,
    redis::keys::{ResourceLookupKey, ServiceKindKey},
};

type ReturnedInstance = (
    String,     // url
    String,     // api key ids
    (u8, bool), // load, accepting jobs
);

#[derive(Debug)]
enum ScriptResult {
    Ok(ReturnedInstance),
    NoInstanceAvailable, // 1
}

impl FromRedisValue for ScriptResult {
    fn from_redis_value(v: redis::Value) -> Result<Self, redis::ParsingError> {
        if let Ok(returned_instance) = ReturnedInstance::from_redis_value(v.clone()) {
            return Ok(ScriptResult::Ok(returned_instance));
        }

        let int_value = i32::from_redis_value(v)?;

        match int_value {
            1 => Ok(ScriptResult::NoInstanceAvailable),
            other => Err(format!("unexpected select_instance script return value: {other}").into()),
        }
    }
}

const SCRIPT: &str = include_str!("select_instance.lua");

/// Select the service that manages the given resource.
///
/// If no service currently manages the resource, the service with the lowest load that is accepting
/// jobs will be selected and assigned to manage the resource.
pub(crate) async fn select_instance(
    client: &redis::Client,
    resource: ServiceResource,
) -> anyhow::Result<(Url, InstanceData)> {
    let script = redis::Script::new(SCRIPT);
    let mut con = client.get_multiplexed_async_connection().await?;

    let kind = match resource {
        ServiceResource::Roomserver(_) => ServiceKind::Roomserver,
        ServiceResource::Recorder(_) => ServiceKind::Recorder,
        ServiceResource::Transcription(_) => ServiceKind::Transcription,
    };

    let result: ScriptResult = script
        .key(ResourceLookupKey { resource })
        .key(ServiceKindKey { kind })
        .arg(resource.to_string())
        .invoke_async(&mut con)
        .await
        .context("Failed to select instance")?;

    let (service_id, api_key_ids, (load, accepting_jobs)) = match result {
        ScriptResult::Ok((service_id, api_key_ids, (load, accepting_jobs))) => {
            (service_id, api_key_ids, (load, accepting_jobs))
        }
        ScriptResult::NoInstanceAvailable => {
            anyhow::bail!("No instance available for resource: {resource:?}")
        }
    };

    let url_str = general_purpose::STANDARD
        .decode(service_id)
        .context("Failed to decode base64 url")?;
    let url_str = String::from_utf8_lossy(url_str.as_ref());
    let url = Url::parse(&url_str).context("Failed to parse url from decoded base64")?;

    let api_key_ids: Vec<ApiKeyId> = serde_json::from_str::<Vec<String>>(&api_key_ids)
        .context("Failed to deserialize api_key_ids")?
        .into_iter()
        .map(ApiKeyId::from)
        .collect();

    let instance_data = InstanceData {
        metrics: Metrics {
            load,
            accepting_jobs,
        },
        api_key_ids,
    };

    Ok((url, instance_data))
}
